<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="Statial">
<title>Statial • Statial</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Statial">
<meta property="og:description" content="Statial">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-dark"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">Statial</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.4.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item">
  <a class="nav-link" href="../articles/Statial.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/SydneyBioX/Statial/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Statial</h1>
                        <h4 data-toc-skip class="author">Farhan
Ameen</h4>
            <address class="author_afil">
      School of Mathematics and Statistics, University of Sydney,
Australia<br><h4 data-toc-skip class="author">Sourish
Iyengar</h4>
            <address class="author_afil">
      School of Mathematics and Statistics, University of Sydney,
Australia<br><h4 data-toc-skip class="author">Alex
Qin</h4>
            <address class="author_afil">
      School of Mathematics and Statistics, University of Sydney,
Australia<br><h4 data-toc-skip class="author">Ellis
Patrick</h4>
            <address class="author_afil">
      Westmead Institute for Medical Research, University of Sydney,
AustraliaSchool of Mathematics and Statistics, University of Sydney,
Australia<br><h4 data-toc-skip class="date">25 October 2023</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/SydneyBioX/Statial/blob/HEAD/vignettes/Statial.Rmd" class="external-link"><code>vignettes/Statial.Rmd</code></a></small>
      <div class="d-none name"><code>Statial.Rmd</code></div>
    </address>
</address>
</address>
</address>
</div>

    
    
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Install the package from Bioconductor</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"Statial"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="load-packages">Load packages<a class="anchor" aria-label="anchor" href="#load-packages"></a>
</h2>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Loading required packages</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://sydneybiox.github.io/Statial">Statial</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ellispatrick.github.io/spicyR/" class="external-link">spicyR</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://sydneybiox.github.io/ClassifyR/" class="external-link">ClassifyR</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ellispatrick.github.io/lisaClust/" class="external-link">lisaClust</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/pharmaverse/ggsurvfit" class="external-link">ggsurvfit</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/therneau/survival" class="external-link">survival</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tibble.tidyverse.org/" class="external-link">tibble</a></span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme_get.html" class="external-link">theme_set</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_classic</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">nCores</span> <span class="op">&lt;-</span> <span class="fl">1</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p>There are over 37 trillion cells in the human body, each taking up
different forms and functions. The behaviour of these cells can be
described by canonical characteristics, but their functions can also
dynamically change based on their environmental context, leading to
cells with diverse states. Understanding changes in cell state and the
interplay between cells is key to understanding their mechanisms of
action and how they contribute to human disease. <code>Statial</code> is
a suite of functions for identifying changes in cell state. This guide
will provide a step-by-step overview of some key functions within
<code>Statial</code>.</p>
</div>
<div class="section level2">
<h2 id="loading-example-data">Loading example data<a class="anchor" aria-label="anchor" href="#loading-example-data"></a>
</h2>
<p>To illustrate the functionality of Statial we will use a multiplexed
ion beam imaging by time-of-flight (MIBI-TOF) dataset profiling tissue
from triple-negative breast cancer patients<span class="math inline">\(^1\)</span> by Keren et al., 2018. This dataset
simultaneously quantifies <em>in situ</em> expression of 36 proteins in
34 immune rich patients. <em>Note:</em> The data contains some
“uninformative” probes and the original cohort included 41 patients.</p>
<p>These images are stored in a <code>SingleCellExperiment</code> object
called <code>kerenSCE</code>. This object contains 57811 cells across 10
images and includes information on cell type and patient survival.</p>
<p><em>Note:</em> The original dataset was reduced down from 41 images
to 10 images for the purposes of this vignette, due to size
restrictions.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load head and neck data</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"kerenSCE"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">kerenSCE</span></span>
<span><span class="co">#&gt; class: SingleCellExperiment </span></span>
<span><span class="co">#&gt; dim: 48 57811 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(1): intensities</span></span>
<span><span class="co">#&gt; rownames(48): Na Si ... Ta Au</span></span>
<span><span class="co">#&gt; rowData names(0):</span></span>
<span><span class="co">#&gt; colnames(57811): 1 2 ... 171281 171282</span></span>
<span><span class="co">#&gt; colData names(8): x y ... Survival_days_capped Censored</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="kontextual-identifying-discrete-changes-in-cell-state">Kontextual: Identifying discrete changes in cell state<a class="anchor" aria-label="anchor" href="#kontextual-identifying-discrete-changes-in-cell-state"></a>
</h2>
<p><code>Kontextual</code> is a method to evaluate the localisation
relationship between two cell types in an image. <code>Kontextual</code>
builds on the L-function by contextualising the relationship between two
cell types in reference to the typical spatial behaviour of a <span class="math inline">\(3^{rd}\)</span> cell type/population. By taking
this approach, <code>Kontextual</code> is invariant to changes in the
window of the image as well as tissue structures which may be
present.</p>
<p>The definitions of cell types and cell states are somewhat ambiguous,
cell types imply well defined groups of cells that serve different roles
from one another, on the other hand cell states imply that cells are a
dynamic entity which cannot be discretised, and thus exist in a
continuum. For the purposes of using <code>Kontextual</code> we treat
cell states as identified clusters of cells, where larger clusters
represent a “parent” cell population, and finer sub-clusters
representing a “child” cell population. For example a CD4 T cell may be
considered a child to a larger parent population of Immune cells.
<code>Kontextual</code> thus aims to see how a child population of cells
deviate from the spatial behaviour of their parent population, and how
that influences the localisation between the child cell state and
another cell state.</p>
<div class="section level3">
<h3 id="cell-type-hierarchy">Cell type hierarchy<a class="anchor" aria-label="anchor" href="#cell-type-hierarchy"></a>
</h3>
<p>A key input for Kontextual is an annotation of cell type hierarchies.
We will need these to organise all the cells present into cell state
populations or clusters, e.g. all the different B cell types are put in
a vector called bcells.</p>
<p>For the purposes of this vignette, these will be manually defined.
Alternatively, you can use the <code>treeKor</code> bioconductor package
<a href="http://www.bioconductor.org/packages/release/bioc/html/treekoR.html" class="external-link">treekoR</a>
to define these hierarchies in a data driven way.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Examine all cell types in image</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">kerenSCE</span><span class="op">$</span><span class="va">cellType</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "Keratin_Tumour" "CD3_Cell"       "B"              "CD4_Cell"      </span></span>
<span><span class="co">#&gt;  [5] "Dc/Mono"        "Unidentified"   "Macrophages"    "CD8_Cell"      </span></span>
<span><span class="co">#&gt;  [9] "other immune"   "Endothelial"    "Mono/Neu"       "Mesenchymal"   </span></span>
<span><span class="co">#&gt; [13] "Neutrophils"    "NK"             "Tumour"         "DC"            </span></span>
<span><span class="co">#&gt; [17] "Tregs"</span></span>
<span></span>
<span><span class="co"># Set up cell populations</span></span>
<span><span class="va">tumour</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Keratin_Tumour"</span>, <span class="st">"Tumour"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">bcells</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"B"</span><span class="op">)</span></span>
<span><span class="va">tcells</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD3_Cell"</span>, <span class="st">"CD4_Cell"</span>, <span class="st">"CD8_Cell"</span>, <span class="st">"Tregs"</span><span class="op">)</span></span>
<span><span class="va">myeloid</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Dc/Mono"</span>, <span class="st">"DC"</span>, <span class="st">"Mono/Neu"</span>, <span class="st">"Macrophages"</span>, <span class="st">"Neutrophils"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">endothelial</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Endothelial"</span><span class="op">)</span></span>
<span><span class="va">mesenchymal</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Mesenchymal"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">tissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">endothelial</span>, <span class="va">mesenchymal</span><span class="op">)</span></span>
<span><span class="va">immune</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">bcells</span>, <span class="va">tcells</span>, <span class="va">myeloid</span>, <span class="st">"NK"</span>, <span class="st">"other immune"</span><span class="op">)</span> <span class="co"># NK = Natural Killer cells</span></span>
<span></span>
<span><span class="va">all</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">tumour</span>, <span class="va">tissue</span>, <span class="va">immune</span>, <span class="st">"Unidentified"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="discrete-cell-state-changes-within-a-single-image">Discrete cell state changes within a single image<a class="anchor" aria-label="anchor" href="#discrete-cell-state-changes-within-a-single-image"></a>
</h3>
<p>Here we examine an image highlighted in the Keren et al. 2018
manuscript where the relationship between two cell types depends on a
parent cell population. In image 6 of the Keren et al. dataset, we can
see that <em>p53+ tumour cells</em> and <em>immune cells</em> are
dispersed. However when the behaviour of <em>p53+ tumour cells</em> are
placed in the context of the spatial behaviour of its broader parent
population <em>tumour cells</em>, <em>p53+ tumour cells</em> and
<em>immune</em> would appear localised.w</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Lets define a new cell type vector</span></span>
<span><span class="va">kerenSCE</span><span class="op">$</span><span class="va">cellTypeNew</span> <span class="op">&lt;-</span> <span class="va">kerenSCE</span><span class="op">$</span><span class="va">cellType</span></span>
<span></span>
<span><span class="co"># Select for all cells that express higher than baseline level of p53</span></span>
<span><span class="va">p53Pos</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">kerenSCE</span><span class="op">)</span><span class="op">[</span><span class="st">"p53"</span>,<span class="op">]</span> <span class="op">&gt;</span> <span class="op">-</span><span class="fl">0.300460</span></span>
<span></span>
<span><span class="co"># Find p53+ tumour cells</span></span>
<span><span class="va">kerenSCE</span><span class="op">$</span><span class="va">cellTypeNew</span><span class="op">[</span><span class="va">kerenSCE</span><span class="op">$</span><span class="va">cellType</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">tumour</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"Tumour"</span></span>
<span><span class="va">kerenSCE</span><span class="op">$</span><span class="va">cellTypeNew</span><span class="op">[</span><span class="va">p53Pos</span> <span class="op">&amp;</span> <span class="va">kerenSCE</span><span class="op">$</span><span class="va">cellType</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">tumour</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"p53_Tumour"</span></span>
<span></span>
<span><span class="co">#Group all immune cells under the name "Immune"</span></span>
<span></span>
<span><span class="va">kerenSCE</span><span class="op">$</span><span class="va">cellTypeNew</span><span class="op">[</span><span class="va">kerenSCE</span><span class="op">$</span><span class="va">cellType</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">immune</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"Immune"</span></span>
<span></span>
<span><span class="co"># Plot image 6</span></span>
<span></span>
<span><span class="va">kerenSCE</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">imageID</span> <span class="op">==</span> <span class="st">"6"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">cellTypeNew</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Immune"</span>, <span class="st">"Tumour"</span>, <span class="st">"p53_Tumour"</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="va">cellTypeNew</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x</span>, y <span class="op">=</span> <span class="va">y</span>, color <span class="op">=</span> <span class="va">cellTypeNew</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_colour_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#505050"</span>, <span class="st">"#64BC46"</span>,<span class="st">"#D6D6D6"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html" class="external-link">guides</a></span><span class="op">(</span>colour <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guide_legend.html" class="external-link">guide_legend</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Cell types"</span>, override.aes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>size<span class="op">=</span><span class="fl">3</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="Statial_files/figure-html/unnamed-chunk-6-1.png" width="700"></p>
<p><code>Kontextual</code> accepts a <code>SingleCellExperiment</code>
object, a single image, or list of images from a
<code>SingleCellExperiment</code> object, which gets passed into the
<code>cells</code> argument. The two cell types which will be evaluated
are specified in the <code>to</code> and <code>from</code> arguments. A
parent population must also be specified in the <code>parent</code>
argument, note the parent cell population must include the
<code>to</code> cell type. The argument <code>r</code> will specify the
radius which the cell relationship will be evaluated on.
<code>Kontextual</code> supports parallel processing, the number of
cores can be specified using the <code>cores</code> argument.
<code>Kontextual</code> can take a single value or multiple values for
each argument and will test all combinations of the arguments
specified.</p>
<p>We can calculate these relationships across all images for a single
radius.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p53_Kontextual</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Kontextual.html">Kontextual</a></span><span class="op">(</span></span>
<span>  cells <span class="op">=</span> <span class="va">kerenSCE</span>,</span>
<span>  r <span class="op">=</span> <span class="fl">50</span>,</span>
<span>  from <span class="op">=</span> <span class="st">"p53_Tumour"</span>,</span>
<span>  to <span class="op">=</span> <span class="st">"Immune"</span>,</span>
<span>  parent <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"p53_Tumour"</span>, <span class="st">"Tumour"</span><span class="op">)</span>,</span>
<span>  cellType <span class="op">=</span> <span class="st">"cellTypeNew"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">p53_Kontextual</span></span>
<span><span class="co">#&gt;    imageID               test original kontextual  r weightQuantile inhom  edge</span></span>
<span><span class="co">#&gt; 1        1 p53_Tumour__Immune       NA         NA 50            0.8  TRUE FALSE</span></span>
<span><span class="co">#&gt; 2       14 p53_Tumour__Immune       NA         NA 50            0.8  TRUE FALSE</span></span>
<span><span class="co">#&gt; 3       18 p53_Tumour__Immune       NA         NA 50            0.8  TRUE FALSE</span></span>
<span><span class="co">#&gt; 4       21 p53_Tumour__Immune       NA         NA 50            0.8  TRUE FALSE</span></span>
<span><span class="co">#&gt; 5       29 p53_Tumour__Immune       NA         NA 50            0.8  TRUE FALSE</span></span>
<span><span class="co">#&gt; 6        3 p53_Tumour__Immune       NA         NA 50            0.8  TRUE FALSE</span></span>
<span><span class="co">#&gt; 7       32 p53_Tumour__Immune       NA         NA 50            0.8  TRUE FALSE</span></span>
<span><span class="co">#&gt; 8       35 p53_Tumour__Immune       NA         NA 50            0.8  TRUE FALSE</span></span>
<span><span class="co">#&gt; 9        5 p53_Tumour__Immune       NA         NA 50            0.8  TRUE FALSE</span></span>
<span><span class="co">#&gt; 10       6 p53_Tumour__Immune       NA         NA 50            0.8  TRUE FALSE</span></span>
<span><span class="co">#&gt;    includeZeroCells window window.length</span></span>
<span><span class="co">#&gt; 1              TRUE convex            NA</span></span>
<span><span class="co">#&gt; 2              TRUE convex            NA</span></span>
<span><span class="co">#&gt; 3              TRUE convex            NA</span></span>
<span><span class="co">#&gt; 4              TRUE convex            NA</span></span>
<span><span class="co">#&gt; 5              TRUE convex            NA</span></span>
<span><span class="co">#&gt; 6              TRUE convex            NA</span></span>
<span><span class="co">#&gt; 7              TRUE convex            NA</span></span>
<span><span class="co">#&gt; 8              TRUE convex            NA</span></span>
<span><span class="co">#&gt; 9              TRUE convex            NA</span></span>
<span><span class="co">#&gt; 10             TRUE convex            NA</span></span></code></pre></div>
<p>The <code>kontextCurve</code> function plots the L-function value and
Kontextual values over a range of radii. If the points lie above the red
line (expected pattern) then localisation is indicated for that radius,
if the points lie below the red line then dispersion is indicated. As
seen in the following plot Kontextual is able to correctly identify
localisation between <code>p53</code> and <code>Immune</code> in the
example image for a certain range of radii. When the radius gets too
large the overall relationship between <code>p53</code> and
<code>Immune</code> looks dispersed. The original L-function is not able
to identify localisation at any value of radii.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">curves</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/kontextCurve.html">kontextCurve</a></span><span class="op">(</span></span>
<span>  cells <span class="op">=</span> <span class="va">kerenSCE</span>,</span>
<span>  from <span class="op">=</span> <span class="st">"p53_Tumour"</span>,</span>
<span>  to <span class="op">=</span> <span class="st">"Immune"</span>,</span>
<span>  parent <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"p53_Tumour"</span>, <span class="st">"Tumour"</span><span class="op">)</span>,</span>
<span>  rs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">10</span>, <span class="fl">510</span>, <span class="fl">100</span><span class="op">)</span>,</span>
<span>  image <span class="op">=</span> <span class="st">"6"</span>,</span>
<span>  cellType <span class="op">=</span> <span class="st">"cellTypeNew"</span>,</span>
<span>  cores <span class="op">=</span> <span class="va">nCores</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/kontextPlot.html">kontextPlot</a></span><span class="op">(</span><span class="va">curves</span><span class="op">)</span></span></code></pre></div>
<p><img src="Statial_files/figure-html/unnamed-chunk-8-1.png" width="700"></p>
<p>Alternatively all pairwise cell relationships and their corresponding
parent in the dataset can be tested. A data frame with all pairwise
combinations can be creating using the <code>parentCombinations</code>
function. This function takes in a vector of all the cells, as well as
all the parent vectors set up earlier. As shown below the output is a
data frame specifying the <code>to</code>, <code>from</code>, and
<code>parent</code> arguments for <code>Kontextual</code>.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get all relationships between cell types and their parents</span></span>
<span><span class="va">parentDf</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/parentCombinations.html">parentCombinations</a></span><span class="op">(</span></span>
<span>  all <span class="op">=</span> <span class="va">all</span>,</span>
<span>  <span class="va">tumour</span>,</span>
<span>  <span class="va">bcells</span>,</span>
<span>  <span class="va">tcells</span>,</span>
<span>  <span class="va">myeloid</span>,</span>
<span>  <span class="va">endothelial</span>,</span>
<span>  <span class="va">mesenchymal</span>,</span>
<span>  <span class="va">tissue</span>,</span>
<span>  <span class="va">immune</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="discrete-cell-state-changes-across-all-images">Discrete cell state changes across all images<a class="anchor" aria-label="anchor" href="#discrete-cell-state-changes-across-all-images"></a>
</h3>
<p>Rather than specifying <code>to</code>, <code>from</code>, and
<code>parent</code> in Kontextual, the output from
<code>parentCombinations</code> can be inputed into
<code>Kontextual</code> using the <code>parentDf</code> argument, to
examine all pairwise relationships in the dataset. This chunk will take
a signficant amount of time to run, for demonstration the results have
been saved and are loaded in.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Running Kontextual on all relationships across all images.</span></span>
<span><span class="va">kerenKontextual</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Kontextual.html">Kontextual</a></span><span class="op">(</span></span>
<span>  cells <span class="op">=</span> <span class="va">kerenSCE</span>,</span>
<span>  parentDf <span class="op">=</span> <span class="va">parentDf</span>,</span>
<span>  r <span class="op">=</span> <span class="fl">50</span>,</span>
<span>  cores <span class="op">=</span> <span class="va">nCores</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"kerenKontextual"</span><span class="op">)</span></span>
<span><span class="va">bigDiff</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="va">kerenKontextual</span><span class="op">$</span><span class="va">original</span> <span class="op">-</span> <span class="va">kerenKontextual</span><span class="op">$</span><span class="va">kontextual</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">kerenKontextual</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">bigDiff</span><span class="op">)</span>,<span class="op">]</span>, <span class="fl">10</span><span class="op">)</span></span>
<span><span class="co">#&gt;      imageID                                test   original kontextual  r</span></span>
<span><span class="co">#&gt; 3393      18       CD3_Cell__Mesenchymal__tcells  42.540392  293.08841 50</span></span>
<span><span class="co">#&gt; 4499       5         Tumour__Endothelial__tumour -22.329210  181.64536 50</span></span>
<span><span class="co">#&gt; 4517      32                   Tumour__B__tumour  71.070831  244.33654 50</span></span>
<span><span class="co">#&gt; 4589       5            Tumour__Mono/Neu__tumour  -8.341033  126.48917 50</span></span>
<span><span class="co">#&gt; 4537      32            Tumour__CD4_Cell__tumour  37.890443  150.58312 50</span></span>
<span><span class="co">#&gt; 1288      35 Macrophages__Keratin_Tumour__immune -12.199943   92.73999 50</span></span>
<span><span class="co">#&gt; 3519       5      CD3_Cell__Unidentified__tcells -27.935002   73.33203 50</span></span>
<span><span class="co">#&gt; 4609       5        Tumour__other immune__tumour   9.069621  106.13098 50</span></span>
<span><span class="co">#&gt; 4509       5         Tumour__Mesenchymal__tumour  -3.104019   93.16538 50</span></span>
<span><span class="co">#&gt; 4539       5            Tumour__CD4_Cell__tumour -15.308151   69.23822 50</span></span>
<span><span class="co">#&gt;      weightQuantile inhom  edge includeZeroCells window window.length</span></span>
<span><span class="co">#&gt; 3393            0.8  TRUE FALSE             TRUE convex            NA</span></span>
<span><span class="co">#&gt; 4499            0.8  TRUE FALSE             TRUE convex            NA</span></span>
<span><span class="co">#&gt; 4517            0.8  TRUE FALSE             TRUE convex            NA</span></span>
<span><span class="co">#&gt; 4589            0.8  TRUE FALSE             TRUE convex            NA</span></span>
<span><span class="co">#&gt; 4537            0.8  TRUE FALSE             TRUE convex            NA</span></span>
<span><span class="co">#&gt; 1288            0.8  TRUE FALSE             TRUE convex            NA</span></span>
<span><span class="co">#&gt; 3519            0.8  TRUE FALSE             TRUE convex            NA</span></span>
<span><span class="co">#&gt; 4609            0.8  TRUE FALSE             TRUE convex            NA</span></span>
<span><span class="co">#&gt; 4509            0.8  TRUE FALSE             TRUE convex            NA</span></span>
<span><span class="co">#&gt; 4539            0.8  TRUE FALSE             TRUE convex            NA</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="associate-discrete-state-changes-with-survival-outcomes">Associate discrete state changes with survival outcomes<a class="anchor" aria-label="anchor" href="#associate-discrete-state-changes-with-survival-outcomes"></a>
</h3>
<p>To examine whether the features obtained from <code>Statial</code>
are associated with patient outcomes or groupings, we can use the
<code>colTest</code> function from <code>SpicyR</code>. To understand if
survival outcomes differ significantly between 2 patient groups, specify
<code>type = "survival"</code> in <code>colTest</code>. Here we examine
which features are most associated with patient survival using the
Kontextual values as an example. To do so, survival data is extracted
from <code>kerenSCE</code> and converted into the survival object
<code>kerenSurv</code>.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># Extracting survival data</span></span>
<span><span class="va">survData</span> <span class="op">=</span> <span class="va">kerenSCE</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">imageID</span>, <span class="va">Survival_days_capped</span>, <span class="va">Censored</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Creating survival vector</span></span>
<span><span class="va">kerenSurv</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/survival/man/Surv.html" class="external-link">Surv</a></span><span class="op">(</span><span class="va">survData</span><span class="op">$</span><span class="va">Survival_days_capped</span>, <span class="va">survData</span><span class="op">$</span><span class="va">Censored</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">kerenSurv</span><span class="op">)</span> <span class="op">=</span> <span class="va">survData</span><span class="op">$</span><span class="va">imageID</span></span></code></pre></div>
<p>In addition to this, the Kontextual results must be converted from a
<code>data.frame</code> to a wide <code>matrix</code>, this can be done
using <code>prepMatrix</code>. Note, to extract the original L-function
values, specify <code>column = "original"</code> in
<code>prepMatrix</code>.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Converting Kontextual result into data matrix</span></span>
<span><span class="va">kontextMat</span> <span class="op">=</span> <span class="fu"><a href="../reference/prepMatrix.html">prepMatrix</a></span><span class="op">(</span><span class="va">kerenKontextual</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Ensuring rownames of kontextMat match up with rownames of the survival vector </span></span>
<span><span class="va">kontextMat</span> <span class="op">=</span> <span class="va">kontextMat</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">kerenSurv</span><span class="op">)</span>, <span class="op">]</span></span>
<span></span>
<span><span class="co"># Replace NAs with 0</span></span>
<span><span class="va">kontextMat</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">kontextMat</span> <span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">0</span></span></code></pre></div>
<p>Finally, both the Kontextual matrix and survival object are passed
into <code>colTest</code>, with <code>type = "survival"</code> to obtain
the survival results.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Running survival analysis</span></span>
<span><span class="va">survivalResults</span> <span class="op">=</span> <span class="fu">spicyR</span><span class="fu">::</span><span class="fu"><a href="https://ellispatrick.github.io/spicyR/reference/colTest.html" class="external-link">colTest</a></span><span class="op">(</span><span class="va">kontextMat</span>, <span class="va">kerenSurv</span>, type <span class="op">=</span> <span class="st">"survival"</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">survivalResults</span><span class="op">)</span></span>
<span><span class="co">#&gt;                                      coef se.coef   pval adjPval</span></span>
<span><span class="co">#&gt; Mesenchymal__Dc/Mono__mesenchymal   1.100   0.330 0.0012    0.56</span></span>
<span><span class="co">#&gt; Mesenchymal__other immune__tissue   0.096   0.048 0.0440    0.62</span></span>
<span><span class="co">#&gt; other immune__Endothelial__myeloid  0.047   0.024 0.0490    0.62</span></span>
<span><span class="co">#&gt; CD4_Cell__Neutrophils__immune       0.110   0.055 0.0500    0.62</span></span>
<span><span class="co">#&gt; CD4_Cell__CD8_Cell__tcells          0.580   0.300 0.0500    0.62</span></span>
<span><span class="co">#&gt; CD4_Cell__Keratin_Tumour__immune   -0.150   0.075 0.0510    0.62</span></span>
<span><span class="co">#&gt;                                                               cluster</span></span>
<span><span class="co">#&gt; Mesenchymal__Dc/Mono__mesenchymal   Mesenchymal__Dc/Mono__mesenchymal</span></span>
<span><span class="co">#&gt; Mesenchymal__other immune__tissue   Mesenchymal__other immune__tissue</span></span>
<span><span class="co">#&gt; other immune__Endothelial__myeloid other immune__Endothelial__myeloid</span></span>
<span><span class="co">#&gt; CD4_Cell__Neutrophils__immune           CD4_Cell__Neutrophils__immune</span></span>
<span><span class="co">#&gt; CD4_Cell__CD8_Cell__tcells                 CD4_Cell__CD8_Cell__tcells</span></span>
<span><span class="co">#&gt; CD4_Cell__Keratin_Tumour__immune     CD4_Cell__Keratin_Tumour__immune</span></span></code></pre></div>
<p>As we can see from the results
<code>Mesenchymal__Dc/Mono__mesenchymal</code> is the most significant
pairwise relationship which contributes to patient survival. That is the
relationship between Mesenchymal cells and macrophage cells, relative to
the parent population of all tissue cells. We can see that there is a
negative coefficient associated with this relationship, which tells us a
decrease in localisation of Mesenchymal and Macrophages leads to poorer
survival outcomes for patients.</p>
<p>The association between
<code>Mesenchymal__Dc/Mono__mesenchymal</code> and survival can also be
visualised on a Kaplan-Meier curve. We must first extract the Kontextual
values of this relationship across all images. Next we determine if
Mesenchymal and Macrophages are relatively attracted or avoiding in each
image, by comparing the Kontextual value in each image to the median
Kontextual value. Finally we plot the Kaplan-Meier curve using the
<code>ggsurvfit</code> package.</p>
<p>As shown below, when Mesenchymal and Macrophages are relatively more
dispersed to one another, patients tend to have worse survival
outcomes.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Selecting most significant relationship</span></span>
<span><span class="va">survRelationship</span> <span class="op">=</span> <span class="va">kontextMat</span><span class="op">[[</span><span class="st">"Mesenchymal__Dc/Mono__mesenchymal"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">survRelationship</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">survRelationship</span> <span class="op">&gt;</span> <span class="fu"><a href="https://rdrr.io/r/stats/median.html" class="external-link">median</a></span><span class="op">(</span><span class="va">survRelationship</span><span class="op">)</span>, <span class="st">"Localised"</span>, <span class="st">"Dispersed"</span><span class="op">)</span></span>
<span>    </span>
<span><span class="co"># Plotting Kaplan-Meier curve</span></span>
<span><span class="fu"><a href="http://www.danieldsjoberg.com/ggsurvfit/reference/survfit2.html" class="external-link">survfit2</a></span><span class="op">(</span><span class="va">kerenSurv</span> <span class="op">~</span> <span class="va">survRelationship</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="http://www.danieldsjoberg.com/ggsurvfit/reference/ggsurvfit.html" class="external-link">ggsurvfit</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="http://www.danieldsjoberg.com/ggsurvfit/reference/add_pvalue.html" class="external-link">add_pvalue</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Mesenchymal__Dc/Mono__mesenchymal"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Statial_files/figure-html/unnamed-chunk-15-1.png" width="480"></p>
</div>
</div>
<div class="section level2">
<h2 id="spatiomark-identifying-continuous-changes-in-cell-state">SpatioMark: Identifying continuous changes in cell state<a class="anchor" aria-label="anchor" href="#spatiomark-identifying-continuous-changes-in-cell-state"></a>
</h2>
<p>Changes in cell states can be analytically framed as the change in
abundance of a gene or protein within a particular cell type. We can use
marker expression to identify and quantify evidence of cell interactions
that catalyse cell state changes. This approach measures how protein
markers in a cell change with spatial proximity and abundance to other
cell types. The methods utilised here will thereby provide a framework
to explore how the dynamic behaviour of cells are altered by the agents
they are surrounded by.</p>
<div class="section level3">
<h3 id="continuous-cell-state-changes-within-a-single-image">Continuous cell state changes within a single image<a class="anchor" aria-label="anchor" href="#continuous-cell-state-changes-within-a-single-image"></a>
</h3>
<p>The first step in analysing these changes is to calculate the spatial
proximity (<code>getDistances</code>) and abundance
(<code>getAbundances</code>) of each cell to every cell type. These
values will then be stored in the <code>reducedDims</code> slot of the
<code>SingleCellExperiment</code> object under the names
<code>distances</code> and <code>abundances</code> respectively.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">kerenSCE</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getDistances.html">getDistances</a></span><span class="op">(</span><span class="va">kerenSCE</span>,</span>
<span>                    maxDist <span class="op">=</span> <span class="fl">200</span>,</span>
<span>                    nCores <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="va">kerenSCE</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getAbundances.html">getAbundances</a></span><span class="op">(</span><span class="va">kerenSCE</span>,</span>
<span>                     r <span class="op">=</span> <span class="fl">200</span>,</span>
<span>                     nCores <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p>First, let’s examine the same effect observed earlier with Kontextual
- the localisation between p53-positive keratin/tumour cells and
macrophages in the context of total keratin/tumour cells for image 6 of
the Keren et al. dataset.</p>
<p>Statial provides two main functions to assess this relationship -
<code>calcStateChanges</code> and <code>plotStateChanges</code>. We can
use <code>calcStateChanges</code> to examine the relationship between 2
cell types for 1 marker in a specific image. In this case, we’re
examining the relationship between keratin/tumour cells
(<code>from = Keratin_Tumour</code>) and macrophages
(<code>to = "Macrophages"</code>) for the marker p53
(<code>marker = "p53"</code>) in <code>image = "6"</code>. We can
appreciate that the <code>fdr</code> statistic for this relationship is
significant, with a negative tvalue, indicating that the expression of
p53 in keratin/tumour cells decreases as distance from macrophages
increases.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">stateChanges</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calcStateChanges.html">calcStateChanges</a></span><span class="op">(</span></span>
<span>  cells <span class="op">=</span> <span class="va">kerenSCE</span>,</span>
<span>  type <span class="op">=</span> <span class="st">"distances"</span>,</span>
<span>  image <span class="op">=</span> <span class="st">"6"</span>,</span>
<span>  from <span class="op">=</span> <span class="st">"Keratin_Tumour"</span>,</span>
<span>  to <span class="op">=</span> <span class="st">"Macrophages"</span>,</span>
<span>  marker <span class="op">=</span> <span class="st">"p53"</span>,</span>
<span>  nCores <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="va">stateChanges</span></span>
<span><span class="co">#&gt;   imageID primaryCellType otherCellType marker         coef      tval</span></span>
<span><span class="co">#&gt; 1       6  Keratin_Tumour   Macrophages    p53 -0.001402178 -7.010113</span></span>
<span><span class="co">#&gt;           pval          fdr</span></span>
<span><span class="co">#&gt; 1 2.868257e-12 2.868257e-12</span></span></code></pre></div>
<p>Statial also provides a convenient function for visualising this
interaction - <code>plotStateChanges</code>. Here, again we can specify
<code>image = 6</code> and our main cell types of interest,
keratin/tumour cells and macrophages, and our marker p53, in the same
format as <code>calcStateChanges</code>.</p>
<p>Through this analysis, we can observe that keratin/tumour cells
closer to a group of macrophages tend to have higher expression of p53,
as observed in the first graph. This relationship is quantified with the
second graph, showing an overall decrease of p53 expression in
keratin/tumour cells as distance to macrophages increase.</p>
<p>These results allow us to essentially arrive at the same result as
Kontextual, which calculated a localisation between p53+ keratin/tumour
cells and macrophages in the wider context of keratin/tumour cells.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotStateChanges.html">plotStateChanges</a></span><span class="op">(</span></span>
<span>  cells <span class="op">=</span> <span class="va">kerenSCE</span>,</span>
<span>  type <span class="op">=</span> <span class="st">"distances"</span>,</span>
<span>  image <span class="op">=</span> <span class="st">"6"</span>,</span>
<span>  from <span class="op">=</span> <span class="st">"Keratin_Tumour"</span>,</span>
<span>  to <span class="op">=</span> <span class="st">"Macrophages"</span>,</span>
<span>  marker <span class="op">=</span> <span class="st">"p53"</span>,</span>
<span>  size <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  shape <span class="op">=</span> <span class="fl">19</span>,</span>
<span>  interactive <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  plotModelFit <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  method <span class="op">=</span> <span class="st">"lm"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p</span></span>
<span><span class="co">#&gt; $image</span></span></code></pre></div>
<p><img src="Statial_files/figure-html/unnamed-chunk-18-1.png" width="700"></p>
<pre><code><span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $scatter</span></span></code></pre>
<p><img src="Statial_files/figure-html/unnamed-chunk-18-2.png" width="700">
## Continuous cell state changes across all images</p>
<p>Beyond looking at single cell-to-cell interactions for a single
image, we can also look at all interactions across all images. The
<code>calcStateChanges</code> function provided by Statial can be
expanded for this exact purpose - by not specifying cell types, a
marker, or an image, <code>calcStateChanges</code> will examine the most
significant correlations between distance and marker expression across
the entire dataset. Here, we’ve filtered out the most significant
interactions to only include those found within image 6 of the Keren et
al. dataset.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">stateChanges</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calcStateChanges.html">calcStateChanges</a></span><span class="op">(</span></span>
<span>  cells <span class="op">=</span> <span class="va">kerenSCE</span>,</span>
<span>  type <span class="op">=</span> <span class="st">"distances"</span>,</span>
<span>  nCores <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  minCells <span class="op">=</span> <span class="fl">100</span><span class="op">)</span></span>
<span></span>
<span><span class="va">stateChanges</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">imageID</span> <span class="op">==</span> <span class="fl">6</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span><span class="co">#&gt;    imageID primaryCellType otherCellType       marker         coef      tval</span></span>
<span><span class="co">#&gt; 1        6  Keratin_Tumour  Unidentified           Na  0.004218419  25.03039</span></span>
<span><span class="co">#&gt; 2        6  Keratin_Tumour   Macrophages  HLA_Class_1 -0.003823497 -24.69629</span></span>
<span><span class="co">#&gt; 3        6  Keratin_Tumour      CD4_Cell  HLA_Class_1 -0.003582774 -23.87797</span></span>
<span><span class="co">#&gt; 4        6  Keratin_Tumour  Unidentified Beta.catenin  0.005893120  23.41953</span></span>
<span><span class="co">#&gt; 5        6  Keratin_Tumour      CD8_Cell  HLA_Class_1 -0.003154544 -23.13804</span></span>
<span><span class="co">#&gt; 6        6  Keratin_Tumour       Dc/Mono  HLA_Class_1 -0.003353834 -22.98944</span></span>
<span><span class="co">#&gt; 7        6  Keratin_Tumour      CD3_Cell  HLA_Class_1 -0.003123446 -22.63197</span></span>
<span><span class="co">#&gt; 8        6  Keratin_Tumour        Tumour  HLA_Class_1  0.003684079  21.94265</span></span>
<span><span class="co">#&gt; 9        6  Keratin_Tumour      CD4_Cell           Fe -0.003457338 -21.43550</span></span>
<span><span class="co">#&gt; 10       6  Keratin_Tumour      CD4_Cell   phospho.S6 -0.002892457 -20.50767</span></span>
<span><span class="co">#&gt;             pval           fdr</span></span>
<span><span class="co">#&gt; 1  6.971648e-127 1.176442e-123</span></span>
<span><span class="co">#&gt; 2  7.814253e-124 1.236215e-120</span></span>
<span><span class="co">#&gt; 3  1.745242e-116 2.208779e-113</span></span>
<span><span class="co">#&gt; 4  1.917245e-112 2.257178e-109</span></span>
<span><span class="co">#&gt; 5  5.444541e-110 5.991836e-107</span></span>
<span><span class="co">#&gt; 6  1.053130e-108 1.110701e-105</span></span>
<span><span class="co">#&gt; 7  1.237988e-105 1.205229e-102</span></span>
<span><span class="co">#&gt; 8  8.188258e-100  7.025803e-97</span></span>
<span><span class="co">#&gt; 9   1.287478e-95  9.727951e-93</span></span>
<span><span class="co">#&gt; 10  3.928912e-88  2.583081e-85</span></span></code></pre></div>
<p>In image 6, the majority of the top 10 most significant interactions
occur between keratin/tumour cells and an immune population, and many of
these interactions appear to involve the HLA class I ligand.</p>
<p>We can examine some of these interactions further with the
<code>plotStateChanges</code> function. Taking a closer examination of
the relationship between macrophages and keratin/tumour HLA class I
expression, the plot below shows us a clear visual correlation - as
macrophage density increases, keratin/tumour cells increase their
expression HLA class I.</p>
<p>Biologically, HLA Class I is a ligand which exists on all nucleated
cells, tasked with presenting internal cell antigens for recognition by
the immune system, marking aberrant cells for destruction by either CD8+
T cells or NK cells.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotStateChanges.html">plotStateChanges</a></span><span class="op">(</span></span>
<span>  cells <span class="op">=</span> <span class="va">kerenSCE</span>,</span>
<span>  type <span class="op">=</span> <span class="st">"distances"</span>,</span>
<span>  image <span class="op">=</span> <span class="st">"6"</span>,</span>
<span>  from <span class="op">=</span> <span class="st">"Keratin_Tumour"</span>,</span>
<span>  to <span class="op">=</span> <span class="st">"Macrophages"</span>,</span>
<span>  marker <span class="op">=</span> <span class="st">"HLA_Class_1"</span>,</span>
<span>  size <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  shape <span class="op">=</span> <span class="fl">19</span>,</span>
<span>  interactive <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  plotModelFit <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  method <span class="op">=</span> <span class="st">"lm"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p</span></span>
<span><span class="co">#&gt; $image</span></span></code></pre></div>
<p><img src="Statial_files/figure-html/unnamed-chunk-20-1.png" width="700"></p>
<pre><code><span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $scatter</span></span></code></pre>
<p><img src="Statial_files/figure-html/unnamed-chunk-20-2.png" width="700"></p>
<p>Next, let’s take a look at the top 10 most significant results across
all images.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">stateChanges</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span><span class="co">#&gt;      imageID primaryCellType otherCellType     marker         coef      tval</span></span>
<span><span class="co">#&gt; 8674      35        CD4_Cell             B       CD20 -0.029185750 -40.57355</span></span>
<span><span class="co">#&gt; 8770      35        CD4_Cell       Dc/Mono       CD20  0.019125946  40.53436</span></span>
<span><span class="co">#&gt; 1819      35               B       Dc/Mono phospho.S6  0.005282065  40.41385</span></span>
<span><span class="co">#&gt; 8779      35        CD4_Cell       Dc/Mono phospho.S6  0.004033218  34.72882</span></span>
<span><span class="co">#&gt; 1813      35               B       Dc/Mono     HLA.DR  0.011120703  34.15344</span></span>
<span><span class="co">#&gt; 1971      35               B  other immune          P  0.011182182  34.14375</span></span>
<span><span class="co">#&gt; 8626      35        CD4_Cell      CD3_Cell       CD20  0.016349492  33.91901</span></span>
<span><span class="co">#&gt; 1816      35               B       Dc/Mono     H3K9ac  0.005096632  33.99856</span></span>
<span><span class="co">#&gt; 2011      35               B  other immune phospho.S6  0.005986586  33.66466</span></span>
<span><span class="co">#&gt; 1818      35               B       Dc/Mono   H3K27me3  0.006980810  33.22740</span></span>
<span><span class="co">#&gt;               pval           fdr</span></span>
<span><span class="co">#&gt; 8674 7.019343e-282 3.553472e-277</span></span>
<span><span class="co">#&gt; 8770 1.891267e-281 4.787176e-277</span></span>
<span><span class="co">#&gt; 1819 5.306590e-278 8.954694e-274</span></span>
<span><span class="co">#&gt; 8779 4.519947e-219 5.720445e-215</span></span>
<span><span class="co">#&gt; 1813 8.401034e-212 8.505879e-208</span></span>
<span><span class="co">#&gt; 1971 1.056403e-211 8.913225e-208</span></span>
<span><span class="co">#&gt; 8626 1.219488e-210 8.819335e-207</span></span>
<span><span class="co">#&gt; 1816 3.266533e-210 2.067062e-206</span></span>
<span><span class="co">#&gt; 2011 8.545691e-207 4.806856e-203</span></span>
<span><span class="co">#&gt; 1818 2.438769e-202 1.234603e-198</span></span></code></pre></div>
<p>Immediately, we can appreciate that a couple of these interactions
are not biologically plausible. One of the most significant interactions
occurs between B cells and CD4 T cells in image 35, where CD4 T cells
are found to increase in CD20 expression when in close proximity to B
cells. Biologically, CD20 is a highly specific ligand for B cells, and
under healthy circumstances are usually not expressed in T cells.</p>
<p>Could this potentially be an artefact of
<code>calcStateChanges</code>? We can examine the image through the
<code>plotStateChanges</code> function, where we indeed observe a strong
increase in CD20 expression in T cells nearby B cell populations.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotStateChanges.html">plotStateChanges</a></span><span class="op">(</span></span>
<span>  cells <span class="op">=</span> <span class="va">kerenSCE</span>,</span>
<span>  type <span class="op">=</span> <span class="st">"distances"</span>,</span>
<span>  image <span class="op">=</span> <span class="st">"35"</span>,</span>
<span>  from <span class="op">=</span> <span class="st">"CD4_Cell"</span>,</span>
<span>  to <span class="op">=</span> <span class="st">"B"</span>,</span>
<span>  marker <span class="op">=</span> <span class="st">"CD20"</span>,</span>
<span>  size <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  shape <span class="op">=</span> <span class="fl">19</span>,</span>
<span>  interactive <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  plotModelFit <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  method <span class="op">=</span> <span class="st">"lm"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p</span></span>
<span><span class="co">#&gt; $image</span></span></code></pre></div>
<p><img src="Statial_files/figure-html/unnamed-chunk-22-1.png" width="700"></p>
<pre><code><span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $scatter</span></span></code></pre>
<p><img src="Statial_files/figure-html/unnamed-chunk-22-2.png" width="700"></p>
<p>So why are T cells expressing CD20? This brings us to a key problem
of cell segmentation - contamination.</p>
</div>
</div>
<div class="section level2">
<h2 id="contamination-lateral-marker-spill-over">Contamination (Lateral marker spill over)<a class="anchor" aria-label="anchor" href="#contamination-lateral-marker-spill-over"></a>
</h2>
<p>Contamination, or lateral marker spill over is an issue that results
in a cell’s marker expressions being wrongly attributed to another
adjacent cell. This issue arises from incorrect segmentation where
components of one cell are wrongly determined as belonging to another
cell. Alternatively, this issue can arise when antibodies used to tag
and measure marker expressions don’t latch on properly to a cell of
interest, thereby resulting in residual markers being wrongly assigned
as belonging to a cell near the intended target cell. It is important
that we either correct or account for this incorrect attribution of
markers in our modelling process. This is critical in understanding
whether significant cell-cell interactions detected are an artefact of
technical measurement errors driven by spill over or are real biological
changes that represent a shift in a cell’s state.</p>
<p>To circumvent this problem, Statial provides a function that predicts
the probability that a cell is any particular cell type -
<code>calcContamination</code>. <code>calcContamination</code> returns a
dataframe of probabilities demarcating the chance of a cell being any
particular cell type. This dataframe is stored under
<code>contaminations</code> in the <code>reducedDim</code> slot of the
<code>SingleCellExperiment</code> object. It also provides the
<code>rfMainCellProb</code> column, which provides the probability that
a cell is indeed the cell type it has been designated. E.g. For a cell
designated as CD8, rfMainCellProb could give a 80% chance that the cell
is indeed CD8, due to contamination.</p>
<p>We can then introduce these probabilities as covariates into our
linear model by setting <code>contamination = TRUE</code> as a parameter
in our <code>calcStateChanges</code> function. However, this is not a
perfect solution for the issue of contamination. As we can see, despite
factoring in contamination into our linear model, the correlation
between B cell density and CD20 expression in CD4 T cells remains one of
the most significant interactions in our model.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">kerenSCE</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calcContamination.html">calcContamination</a></span><span class="op">(</span><span class="va">kerenSCE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">stateChangesCorrected</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calcStateChanges.html">calcStateChanges</a></span><span class="op">(</span></span>
<span>  cells <span class="op">=</span> <span class="va">kerenSCE</span>,</span>
<span>  type <span class="op">=</span> <span class="st">"distances"</span>,</span>
<span>  nCores <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  minCells <span class="op">=</span> <span class="fl">100</span>,</span>
<span>  contamination <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">stateChangesCorrected</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></span>
<span><span class="co">#&gt;       imageID primaryCellType otherCellType      marker         coef      tval</span></span>
<span><span class="co">#&gt; 8674       35        CD4_Cell             B        CD20 -0.024786419 -34.95527</span></span>
<span><span class="co">#&gt; 8770       35        CD4_Cell       Dc/Mono        CD20  0.015872636  33.46048</span></span>
<span><span class="co">#&gt; 8779       35        CD4_Cell       Dc/Mono  phospho.S6  0.003580507  29.44495</span></span>
<span><span class="co">#&gt; 1819       35               B       Dc/Mono  phospho.S6  0.004240940  29.40981</span></span>
<span><span class="co">#&gt; 8626       35        CD4_Cell      CD3_Cell        CD20  0.013479231  29.16784</span></span>
<span><span class="co">#&gt; 8629       35        CD4_Cell      CD3_Cell      HLA.DR  0.010217422  28.71325</span></span>
<span><span class="co">#&gt; 29188       3  Keratin_Tumour            DC          Ca -0.013671664 -28.87189</span></span>
<span><span class="co">#&gt; 1669       35               B      CD3_Cell      HLA.DR  0.008906451  25.52653</span></span>
<span><span class="co">#&gt; 27641      21  Keratin_Tumour            DC Pan.Keratin -0.006003833 -24.85754</span></span>
<span><span class="co">#&gt; 1813       35               B       Dc/Mono      HLA.DR  0.008809238  25.09780</span></span>
<span><span class="co">#&gt; 8763       35        CD4_Cell       Dc/Mono      CSF.1R  0.008552018  24.88996</span></span>
<span><span class="co">#&gt; 31825       6  Keratin_Tumour  Unidentified          Na  0.004216952  24.62453</span></span>
<span><span class="co">#&gt; 8635       35        CD4_Cell      CD3_Cell  phospho.S6  0.002914333  24.55304</span></span>
<span><span class="co">#&gt; 2011       35               B  other immune  phospho.S6  0.004495837  24.04748</span></span>
<span><span class="co">#&gt; 1675       35               B      CD3_Cell  phospho.S6  0.003565036  23.67518</span></span>
<span><span class="co">#&gt; 31918       6  Keratin_Tumour   Macrophages HLA_Class_1 -0.003441291 -22.78547</span></span>
<span><span class="co">#&gt; 1816       35               B       Dc/Mono      H3K9ac  0.003725499  22.92998</span></span>
<span><span class="co">#&gt; 2008       35               B  other immune      H3K9ac  0.004607261  22.81682</span></span>
<span><span class="co">#&gt; 31774       6  Keratin_Tumour      CD4_Cell HLA_Class_1 -0.003277158 -22.54297</span></span>
<span><span class="co">#&gt; 29186       3  Keratin_Tumour            DC          Si -0.005542163 -22.94915</span></span>
<span><span class="co">#&gt;                pval           fdr</span></span>
<span><span class="co">#&gt; 8674  3.141399e-221 1.590302e-216</span></span>
<span><span class="co">#&gt; 8770  1.023792e-205 2.591422e-201</span></span>
<span><span class="co">#&gt; 8779  1.532688e-165 2.586360e-161</span></span>
<span><span class="co">#&gt; 1819  2.069706e-164 2.619420e-160</span></span>
<span><span class="co">#&gt; 8626  7.336926e-163 7.428491e-159</span></span>
<span><span class="co">#&gt; 8629  1.711503e-158 1.444052e-154</span></span>
<span><span class="co">#&gt; 29188 5.168071e-156 3.737549e-152</span></span>
<span><span class="co">#&gt; 1669  1.802832e-128 1.140832e-124</span></span>
<span><span class="co">#&gt; 27641 2.879473e-128 1.619671e-124</span></span>
<span><span class="co">#&gt; 1813  1.107925e-124 5.608757e-121</span></span>
<span><span class="co">#&gt; 8763  2.627293e-123 1.209128e-119</span></span>
<span><span class="co">#&gt; 31825 3.892314e-123 1.642037e-119</span></span>
<span><span class="co">#&gt; 8635  2.371224e-120 9.233913e-117</span></span>
<span><span class="co">#&gt; 2011  1.440258e-115 5.207974e-112</span></span>
<span><span class="co">#&gt; 1675  2.138478e-112 7.217220e-109</span></span>
<span><span class="co">#&gt; 31918 6.507296e-107 2.058908e-103</span></span>
<span><span class="co">#&gt; 1816  3.819083e-106 1.137278e-102</span></span>
<span><span class="co">#&gt; 2008  3.310591e-105 9.310853e-102</span></span>
<span><span class="co">#&gt; 31774 7.675714e-105 2.045134e-101</span></span>
<span><span class="co">#&gt; 29186 8.200541e-105 2.075721e-101</span></span></code></pre></div>
<p>However, this does not mean factoring in contamination into our
linear model was ineffective.</p>
<p>Whilst our correction attempts do not rectify every relationship
which arises due to contamination, we show that a significant portion of
these relationships are rectified. We can show this by plotting a ROC
curve of true positives against false positives. In general, cell type
specific markers such as CD4, CD8, and CD20 should not change in cells
they are not specific to. Therefore, relationships detected to be
significant involving these cell type markers are likely false positives
and will be treated as such for the purposes of evaluation. Meanwhile,
cell state markers are predominantly likely to be true positives.</p>
<p>Plotting the relationship between false positives and true positives,
we’d expect the contamination correction to be greatest in the
relationships with the top 100 lowest p values, where we indeed see more
true positives than false positives with contamination correction.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cellTypeMarkers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD3"</span>, <span class="st">"CD4"</span>, <span class="st">"CD8"</span>, <span class="st">"CD56"</span>, <span class="st">"CD11c"</span>, <span class="st">"CD68"</span>, <span class="st">"CD45"</span>, <span class="st">"CD20"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">values</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"blue"</span>, <span class="st">"red"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">values</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"None"</span>, <span class="st">"Corrected"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">rbind</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>TP <span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/cumsum.html" class="external-link">cumsum</a></span><span class="op">(</span><span class="va">stateChanges</span><span class="op">$</span><span class="va">marker</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">cellTypeMarkers</span><span class="op">)</span>, FP <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/cumsum.html" class="external-link">cumsum</a></span><span class="op">(</span><span class="op">!</span><span class="va">stateChanges</span><span class="op">$</span><span class="va">marker</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">cellTypeMarkers</span><span class="op">)</span>, type <span class="op">=</span> <span class="st">"None"</span><span class="op">)</span>,</span>
<span>            <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>TP <span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/cumsum.html" class="external-link">cumsum</a></span><span class="op">(</span><span class="va">stateChangesCorrected</span><span class="op">$</span><span class="va">marker</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">cellTypeMarkers</span><span class="op">)</span>, FP <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/cumsum.html" class="external-link">cumsum</a></span><span class="op">(</span><span class="op">!</span><span class="va">stateChangesCorrected</span><span class="op">$</span><span class="va">marker</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">cellTypeMarkers</span><span class="op">)</span>, type <span class="op">=</span> <span class="st">"Corrected"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">TP</span>, y <span class="op">=</span> <span class="va">FP</span>, colour <span class="op">=</span> <span class="va">type</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html" class="external-link">geom_line</a></span><span class="op">(</span><span class="op">)</span><span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>y <span class="op">=</span> <span class="st">"Cell state marker"</span>, x <span class="op">=</span> <span class="st">"Cell type marker"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_colour_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">values</span><span class="op">)</span></span></code></pre></div>
<p><img src="Statial_files/figure-html/unnamed-chunk-24-1.png" width="700"></p>
<p>Here, we zoom in on the ROC curve where the top 100 lowest p values
occur, where we indeed see more true positives than false positives with
contamination correction.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">TP</span>, y <span class="op">=</span> <span class="va">FP</span>, colour <span class="op">=</span> <span class="va">type</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html" class="external-link">geom_line</a></span><span class="op">(</span><span class="op">)</span><span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html" class="external-link">xlim</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">100</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html" class="external-link">ylim</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1000</span><span class="op">)</span><span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>y <span class="op">=</span> <span class="st">"Cell state marker"</span>, x <span class="op">=</span> <span class="st">"Cell type marker"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_colour_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">values</span><span class="op">)</span></span></code></pre></div>
<p><img src="Statial_files/figure-html/unnamed-chunk-25-1.png" width="700"></p>
<div class="section level3">
<h3 id="associate-continuous-state-changes-with-survival-outcomes">Associate continuous state changes with survival outcomes<a class="anchor" aria-label="anchor" href="#associate-continuous-state-changes-with-survival-outcomes"></a>
</h3>
<p>Similiar to <code>Kontextual</code>, we can run a similar survival
analysis using our state changes results. Here, <code>prepMatrix</code>
extracts the coefficients, or the <code>coef</code> column of
<code>stateChanges</code> by default. To use the t values instead,
specify <code>column = "tval"</code> in the <code>prepMatrix</code>
function.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Preparing features for Statial</span></span>
<span><span class="va">stateMat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/prepMatrix.html">prepMatrix</a></span><span class="op">(</span><span class="va">stateChanges</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Ensuring rownames of stateMat match up with rownames of the survival vector</span></span>
<span><span class="va">stateMat</span> <span class="op">&lt;-</span> <span class="va">stateMat</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">kerenSurv</span><span class="op">)</span>, <span class="op">]</span></span>
<span></span>
<span><span class="co"># Remove some very small values</span></span>
<span><span class="va">stateMat</span> <span class="op">&lt;-</span> <span class="va">stateMat</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">colMeans</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">stateMat</span><span class="op">)</span><span class="op">&gt;</span><span class="fl">0.0001</span><span class="op">)</span><span class="op">&gt;</span><span class="fl">.8</span><span class="op">]</span></span>
<span></span>
<span><span class="va">survivalResults</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ellispatrick.github.io/spicyR/reference/colTest.html" class="external-link">colTest</a></span><span class="op">(</span><span class="va">stateMat</span>, <span class="va">kerenSurv</span>, type <span class="op">=</span> <span class="st">"survival"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">survivalResults</span><span class="op">)</span></span>
<span><span class="co">#&gt;                                       coef se.coef  pval adjPval</span></span>
<span><span class="co">#&gt; Keratin_Tumour__Keratin_Tumour__EGFR  -390     190 0.037    0.74</span></span>
<span><span class="co">#&gt; Macrophages__Mesenchymal__H3K9ac     -1400     690 0.038    0.74</span></span>
<span><span class="co">#&gt; Macrophages__CD8_Cell__CD45           -880     440 0.045    0.74</span></span>
<span><span class="co">#&gt; Macrophages__CD8_Cell__IDO           -1500     790 0.050    0.74</span></span>
<span><span class="co">#&gt; Macrophages__Mesenchymal__IDO          580     300 0.051    0.74</span></span>
<span><span class="co">#&gt; Keratin_Tumour__CD4_Cell__Keratin6    1600     830 0.051    0.74</span></span>
<span><span class="co">#&gt;                                                                   cluster</span></span>
<span><span class="co">#&gt; Keratin_Tumour__Keratin_Tumour__EGFR Keratin_Tumour__Keratin_Tumour__EGFR</span></span>
<span><span class="co">#&gt; Macrophages__Mesenchymal__H3K9ac         Macrophages__Mesenchymal__H3K9ac</span></span>
<span><span class="co">#&gt; Macrophages__CD8_Cell__CD45                   Macrophages__CD8_Cell__CD45</span></span>
<span><span class="co">#&gt; Macrophages__CD8_Cell__IDO                     Macrophages__CD8_Cell__IDO</span></span>
<span><span class="co">#&gt; Macrophages__Mesenchymal__IDO               Macrophages__Mesenchymal__IDO</span></span>
<span><span class="co">#&gt; Keratin_Tumour__CD4_Cell__Keratin6     Keratin_Tumour__CD4_Cell__Keratin6</span></span></code></pre></div>
<p>For our state changes results,
<code>Keratin_Tumour__CD4_Cell__Keratin6</code> is the most significant
pairwise relationship which contributes to patient survival. That is,
the relationship between HLA class I expression in keratin/tumour cells
and their spatial proximity to mesenchymal cells. As there is a negative
coeffcient associated with this relationship, which tells us that higher
HLA class I expression in keratin/tumour cells nearby mesenchymal cell
populations lead to poorer survival outcomes for patients.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Selecting the most significant relationship</span></span>
<span><span class="va">survRelationship</span> <span class="op">=</span> <span class="va">stateMat</span><span class="op">[[</span><span class="st">"Keratin_Tumour__CD4_Cell__Keratin6"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">survRelationship</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">survRelationship</span> <span class="op">&gt;</span> <span class="fu"><a href="https://rdrr.io/r/stats/median.html" class="external-link">median</a></span><span class="op">(</span><span class="va">survRelationship</span><span class="op">)</span>, <span class="st">"Higher expression in close cells"</span>, <span class="st">"Lower expression in close cells"</span><span class="op">)</span></span>
<span>    </span>
<span><span class="co"># Plotting Kaplan-Meier curve</span></span>
<span><span class="fu"><a href="http://www.danieldsjoberg.com/ggsurvfit/reference/survfit2.html" class="external-link">survfit2</a></span><span class="op">(</span><span class="va">kerenSurv</span> <span class="op">~</span> <span class="va">survRelationship</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="http://www.danieldsjoberg.com/ggsurvfit/reference/ggsurvfit.html" class="external-link">ggsurvfit</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="http://www.danieldsjoberg.com/ggsurvfit/reference/add_pvalue.html" class="external-link">add_pvalue</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Keratin_Tumour__CD4_Cell__Keratin6"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Statial_files/figure-html/unnamed-chunk-27-1.png" width="480"></p>
</div>
</div>
<div class="section level2">
<h2 id="region-analysis-using-lisaclust">Region analysis using lisaClust<a class="anchor" aria-label="anchor" href="#region-analysis-using-lisaclust"></a>
</h2>
<p>Next we can cluster areas with similar spatial interactions to
identify regions using lisaClust. Here we set <code>k = 5</code> to
identify 5 regions.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">51773</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Preparing features for lisaClust</span></span>
<span><span class="va">kerenSCE</span> <span class="op">&lt;-</span> <span class="fu">lisaClust</span><span class="fu">::</span><span class="fu"><a href="https://github.com/ellispatrick/lisaClust/reference/lisaClust.html" class="external-link">lisaClust</a></span><span class="op">(</span><span class="va">kerenSCE</span>, k <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span></code></pre></div>
<p>The regions identified by licaClust can be visualised using the
<code>hatchingPlot</code> function.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Use hatching to visualise regions and cell types.</span></span>
<span><span class="fu">lisaClust</span><span class="fu">::</span><span class="fu"><a href="https://github.com/ellispatrick/lisaClust/reference/hatchingPlot.html" class="external-link">hatchingPlot</a></span><span class="op">(</span><span class="va">kerenSCE</span>,</span>
<span>  useImages <span class="op">=</span> <span class="st">"5"</span>,</span>
<span>  line.spacing <span class="op">=</span> <span class="fl">41</span>, <span class="co"># spacing of lines</span></span>
<span>  nbp <span class="op">=</span> <span class="fl">100</span> <span class="co"># smoothness of lines</span></span>
<span><span class="op">)</span> </span></code></pre></div>
<p><img src="Statial_files/figure-html/unnamed-chunk-29-1.png" width="624"></p>
</div>
<div class="section level2">
<h2 id="marker-means">Marker Means<a class="anchor" aria-label="anchor" href="#marker-means"></a>
</h2>
<p><code>Statial</code> provides functionality to identify the average
marker expression of a given cell type in a given region, using the
<code>getMarkerMeans</code> function. Similar to the analysis above,
these features can also be used for survival analysis.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">cellTypeRegionMeans</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getMarkerMeans.html">getMarkerMeans</a></span><span class="op">(</span><span class="va">kerenSCE</span>,</span>
<span>                              imageID <span class="op">=</span> <span class="st">"imageID"</span>,</span>
<span>                              cellType <span class="op">=</span> <span class="st">"cellType"</span>,</span>
<span>                              region <span class="op">=</span> <span class="st">"region"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">survivalResults</span> <span class="op">=</span> <span class="fu"><a href="https://ellispatrick.github.io/spicyR/reference/colTest.html" class="external-link">colTest</a></span><span class="op">(</span><span class="va">cellTypeRegionMeans</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">kerenSurv</span><span class="op">)</span>,<span class="op">]</span>, <span class="va">kerenSurv</span>, type <span class="op">=</span> <span class="st">"survival"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">survivalResults</span><span class="op">)</span></span>
<span><span class="co">#&gt;                               coef se.coef    pval adjPval</span></span>
<span><span class="co">#&gt; PD.L1__CD3_Cell__region_1     2000    85.0 0.0e+00  0.0000</span></span>
<span><span class="co">#&gt; CD31__CD8_Cell__region_3       190     5.4 0.0e+00  0.0000</span></span>
<span><span class="co">#&gt; IDO__Keratin_Tumour__region_5   20     2.0 0.0e+00  0.0000</span></span>
<span><span class="co">#&gt; IDO__Macrophages__region_5      47    10.0 2.5e-06  0.0024</span></span>
<span><span class="co">#&gt; CD138__B__region_5            -180    70.0 1.0e-02  0.8100</span></span>
<span><span class="co">#&gt; HLA_Class_1__B__region_5        33    13.0 1.2e-02  0.8100</span></span>
<span><span class="co">#&gt;                                                     cluster</span></span>
<span><span class="co">#&gt; PD.L1__CD3_Cell__region_1         PD.L1__CD3_Cell__region_1</span></span>
<span><span class="co">#&gt; CD31__CD8_Cell__region_3           CD31__CD8_Cell__region_3</span></span>
<span><span class="co">#&gt; IDO__Keratin_Tumour__region_5 IDO__Keratin_Tumour__region_5</span></span>
<span><span class="co">#&gt; IDO__Macrophages__region_5       IDO__Macrophages__region_5</span></span>
<span><span class="co">#&gt; CD138__B__region_5                       CD138__B__region_5</span></span>
<span><span class="co">#&gt; HLA_Class_1__B__region_5           HLA_Class_1__B__region_5</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="patient-classification">Patient classification<a class="anchor" aria-label="anchor" href="#patient-classification"></a>
</h2>
<p>Finally we demonstrate how we can use <code>ClassifyR</code> to
perform patient classification with the features generated from
<code>Statial</code>. In addition to the kontextual, state changes, and
marker means values, we also calculate cell type proportions and region
proportions using the <code>getProp</code> function in
<code>spicyR</code>. Here we perform 3 fold cross validation with 10
repeats, using a CoxPH model for survival classification, feature
selection is also performed by selecting the top 10 features per fold
using a CoxPH model.</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># Calculate cell type and region proportions</span></span>
<span><span class="va">cellTypeProp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ellispatrick.github.io/spicyR/reference/getProp.html" class="external-link">getProp</a></span><span class="op">(</span><span class="va">kerenSCE</span>, </span>
<span>                       feature <span class="op">=</span> <span class="st">"cellType"</span>,</span>
<span>                       imageID <span class="op">=</span> <span class="st">"imageID"</span><span class="op">)</span></span>
<span><span class="va">regionProp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ellispatrick.github.io/spicyR/reference/getProp.html" class="external-link">getProp</a></span><span class="op">(</span><span class="va">kerenSCE</span>, </span>
<span>                       feature <span class="op">=</span> <span class="st">"region"</span>,</span>
<span>                       imageID <span class="op">=</span> <span class="st">"imageID"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Combine all the features into a list for classification </span></span>
<span><span class="va">featureList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>states <span class="op">=</span> <span class="va">stateMat</span>, </span>
<span>                     kontextual <span class="op">=</span> <span class="va">kontextMat</span>,</span>
<span>                     regionMarkerMeans <span class="op">=</span> <span class="va">cellTypeRegionMeans</span>,</span>
<span>                     cellTypeProp <span class="op">=</span> <span class="va">cellTypeProp</span>,</span>
<span>                     regionProp <span class="op">=</span> <span class="va">regionProp</span></span>
<span>                    <span class="op">)</span></span>
<span></span>
<span><span class="co"># Ensure the rownames of the features match the order of the survival vector</span></span>
<span><span class="va">featureList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">featureList</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="va">x</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">kerenSurv</span><span class="op">)</span>,<span class="op">]</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">51773</span><span class="op">)</span></span>
<span></span>
<span><span class="va">kerenCV</span> <span class="op">=</span> <span class="fu"><a href="https://sydneybiox.github.io/ClassifyR/reference/crossValidate.html" class="external-link">crossValidate</a></span><span class="op">(</span></span>
<span>  measurements <span class="op">=</span> <span class="va">featureList</span>,</span>
<span>  outcome <span class="op">=</span> <span class="va">kerenSurv</span>,</span>
<span>  classifier <span class="op">=</span> <span class="st">"CoxPH"</span>,</span>
<span>  selectionMethod  <span class="op">=</span> <span class="st">"CoxPH"</span>,</span>
<span>  nFolds <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  nFeatures <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  nRepeats <span class="op">=</span> <span class="fl">20</span>,</span>
<span>  nCores <span class="op">=</span> <span class="fl">1</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p>Here, we use the <code>performancePlot</code> function to assess the
C-index from each repeat of the 3-fold cross-validation. We can see the
resulting C-indexes are very variable due to the dataset only containing
10 images.</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Calculate AUC for each cross-validation repeat and plot.</span></span>
<span><span class="fu"><a href="https://sydneybiox.github.io/ClassifyR/reference/performancePlot.html" class="external-link">performancePlot</a></span><span class="op">(</span><span class="va">kerenCV</span>,</span>
<span>  characteristicsList <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Assay Name"</span><span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">45</span>, vjust <span class="op">=</span> <span class="fl">1</span>, hjust<span class="op">=</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="Statial_files/figure-html/unnamed-chunk-31-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<p>Keren, L., Bosse, M., Marquez, D., Angoshtari, R., Jain, S., Varma,
S., Yang, S. R., Kurian, A., Van Valen, D., West, R., Bendall, S. C.,
&amp; Angelo, M. (2018). A Structured Tumor-Immune Microenvironment in
Triple Negative Breast Cancer Revealed by Multiplexed Ion Beam Imaging.
Cell, 174(6), 1373-1387.e1319. (<a href="https://doi.org/10.1016/j.cell.2018.08.039" class="external-link">DOI</a>)</p>
</div>
<div class="section level2">
<h2 id="sessioninfo">sessionInfo<a class="anchor" aria-label="anchor" href="#sessioninfo"></a>
</h2>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.3.1 (2023-06-16)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">#&gt; Running under: EndeavourOS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/libblas.so.3.11.0 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/liblapack.so.3.11.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=en_AU.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">#&gt;  [3] LC_TIME=en_AU.UTF-8        LC_COLLATE=en_AU.UTF-8    </span></span>
<span><span class="co">#&gt;  [5] LC_MONETARY=en_AU.UTF-8    LC_MESSAGES=en_AU.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=en_AU.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">#&gt;  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">#&gt; [11] LC_MEASUREMENT=en_AU.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: Australia/Sydney</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">#&gt; [8] base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] tibble_3.2.1                ggsurvfit_0.3.1            </span></span>
<span><span class="co">#&gt;  [3] ggplot2_3.4.4               SingleCellExperiment_1.22.0</span></span>
<span><span class="co">#&gt;  [5] dplyr_1.1.3                 lisaClust_1.8.1            </span></span>
<span><span class="co">#&gt;  [7] ClassifyR_3.4.11            survival_3.5-7             </span></span>
<span><span class="co">#&gt;  [9] BiocParallel_1.34.2         MultiAssayExperiment_1.26.0</span></span>
<span><span class="co">#&gt; [11] SummarizedExperiment_1.30.2 Biobase_2.60.0             </span></span>
<span><span class="co">#&gt; [13] GenomicRanges_1.52.1        GenomeInfoDb_1.36.4        </span></span>
<span><span class="co">#&gt; [15] IRanges_2.34.1              MatrixGenerics_1.12.3      </span></span>
<span><span class="co">#&gt; [17] matrixStats_1.0.0           S4Vectors_0.38.2           </span></span>
<span><span class="co">#&gt; [19] BiocGenerics_0.46.0         generics_0.1.3             </span></span>
<span><span class="co">#&gt; [21] spicyR_1.12.2               Statial_1.2.4              </span></span>
<span><span class="co">#&gt; [23] BiocStyle_2.28.1           </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] RColorBrewer_1.1-3        jsonlite_1.8.7           </span></span>
<span><span class="co">#&gt;   [3] magrittr_2.0.3            spatstat.utils_3.0-3     </span></span>
<span><span class="co">#&gt;   [5] magick_2.7.5              nloptr_2.0.3             </span></span>
<span><span class="co">#&gt;   [7] farver_2.1.1              rmarkdown_2.24           </span></span>
<span><span class="co">#&gt;   [9] fs_1.6.3                  zlibbioc_1.46.0          </span></span>
<span><span class="co">#&gt;  [11] ragg_1.2.5                vctrs_0.6.4              </span></span>
<span><span class="co">#&gt;  [13] spatstat.explore_3.2-3    memoise_2.0.1            </span></span>
<span><span class="co">#&gt;  [15] minqa_1.2.6               DelayedMatrixStats_1.22.6</span></span>
<span><span class="co">#&gt;  [17] RCurl_1.98-1.12           rstatix_0.7.2            </span></span>
<span><span class="co">#&gt;  [19] htmltools_0.5.6           S4Arrays_1.0.6           </span></span>
<span><span class="co">#&gt;  [21] curl_5.0.2                broom_1.0.5              </span></span>
<span><span class="co">#&gt;  [23] Rhdf5lib_1.22.1           rhdf5_2.44.0             </span></span>
<span><span class="co">#&gt;  [25] sass_0.4.7                bslib_0.5.1              </span></span>
<span><span class="co">#&gt;  [27] htmlwidgets_1.6.2         desc_1.4.2               </span></span>
<span><span class="co">#&gt;  [29] plyr_1.8.9                plotly_4.10.2            </span></span>
<span><span class="co">#&gt;  [31] cachem_1.0.8              lifecycle_1.0.3          </span></span>
<span><span class="co">#&gt;  [33] pkgconfig_2.0.3           Matrix_1.6-1             </span></span>
<span><span class="co">#&gt;  [35] R6_2.5.1                  fastmap_1.1.1            </span></span>
<span><span class="co">#&gt;  [37] GenomeInfoDbData_1.2.10   numDeriv_2016.8-1.1      </span></span>
<span><span class="co">#&gt;  [39] digest_0.6.33             colorspace_2.1-0         </span></span>
<span><span class="co">#&gt;  [41] tensor_1.5                rprojroot_2.0.3          </span></span>
<span><span class="co">#&gt;  [43] dqrng_0.3.1               textshaping_0.3.6        </span></span>
<span><span class="co">#&gt;  [45] ggpubr_0.6.0              beachmat_2.16.0          </span></span>
<span><span class="co">#&gt;  [47] labeling_0.4.3            spatstat.sparse_3.0-2    </span></span>
<span><span class="co">#&gt;  [49] fansi_1.0.5               mgcv_1.8-42              </span></span>
<span><span class="co">#&gt;  [51] httr_1.4.7                polyclip_1.10-4          </span></span>
<span><span class="co">#&gt;  [53] abind_1.4-5               compiler_4.3.1           </span></span>
<span><span class="co">#&gt;  [55] withr_2.5.1               backports_1.4.1          </span></span>
<span><span class="co">#&gt;  [57] carData_3.0-5             HDF5Array_1.28.1         </span></span>
<span><span class="co">#&gt;  [59] ggforce_0.4.1             R.utils_2.12.2           </span></span>
<span><span class="co">#&gt;  [61] ggsignif_0.6.4            MASS_7.3-60              </span></span>
<span><span class="co">#&gt;  [63] concaveman_1.1.0          DelayedArray_0.26.7      </span></span>
<span><span class="co">#&gt;  [65] rjson_0.2.21              tools_4.3.1              </span></span>
<span><span class="co">#&gt;  [67] ranger_0.15.1             goftest_1.2-3            </span></span>
<span><span class="co">#&gt;  [69] R.oo_1.25.0               glue_1.6.2               </span></span>
<span><span class="co">#&gt;  [71] nlme_3.1-162              rhdf5filters_1.12.1      </span></span>
<span><span class="co">#&gt;  [73] grid_4.3.1                reshape2_1.4.4           </span></span>
<span><span class="co">#&gt;  [75] gtable_0.3.4              spatstat.data_3.0-1      </span></span>
<span><span class="co">#&gt;  [77] class_7.3-22              R.methodsS3_1.8.2        </span></span>
<span><span class="co">#&gt;  [79] tidyr_1.3.0               data.table_1.14.8        </span></span>
<span><span class="co">#&gt;  [81] car_3.1-2                 utf8_1.2.4               </span></span>
<span><span class="co">#&gt;  [83] XVector_0.40.0            spatstat.geom_3.2-5      </span></span>
<span><span class="co">#&gt;  [85] pillar_1.9.0              stringr_1.5.0            </span></span>
<span><span class="co">#&gt;  [87] limma_3.56.2              splines_4.3.1            </span></span>
<span><span class="co">#&gt;  [89] tweenr_2.0.2              lattice_0.21-8           </span></span>
<span><span class="co">#&gt;  [91] deldir_1.0-9              tidyselect_1.2.0         </span></span>
<span><span class="co">#&gt;  [93] locfit_1.5-9.8            scuttle_1.10.3           </span></span>
<span><span class="co">#&gt;  [95] knitr_1.44                V8_4.3.3                 </span></span>
<span><span class="co">#&gt;  [97] bookdown_0.35             edgeR_3.42.4             </span></span>
<span><span class="co">#&gt;  [99] xfun_0.40                 DropletUtils_1.20.0      </span></span>
<span><span class="co">#&gt; [101] pheatmap_1.0.12           fftwtools_0.9-11         </span></span>
<span><span class="co">#&gt; [103] scam_1.2-14               stringi_1.7.12           </span></span>
<span><span class="co">#&gt; [105] lazyeval_0.2.2            yaml_2.3.7               </span></span>
<span><span class="co">#&gt; [107] boot_1.3-28.1             evaluate_0.21            </span></span>
<span><span class="co">#&gt; [109] codetools_0.2-19          BiocManager_1.30.22      </span></span>
<span><span class="co">#&gt; [111] cli_3.6.1                 systemfonts_1.0.4        </span></span>
<span><span class="co">#&gt; [113] munsell_0.5.0             jquerylib_0.1.4          </span></span>
<span><span class="co">#&gt; [115] Rcpp_1.0.11               spatstat.random_3.1-6    </span></span>
<span><span class="co">#&gt; [117] parallel_4.3.1            pkgdown_2.0.7            </span></span>
<span><span class="co">#&gt; [119] sparseMatrixStats_1.12.2  bitops_1.0-7             </span></span>
<span><span class="co">#&gt; [121] lme4_1.1-34               SpatialExperiment_1.10.0 </span></span>
<span><span class="co">#&gt; [123] viridisLite_0.4.2         lmerTest_3.1-3           </span></span>
<span><span class="co">#&gt; [125] scales_1.2.1              purrr_1.0.2              </span></span>
<span><span class="co">#&gt; [127] crayon_1.5.2              rlang_1.1.1</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Farhan Ameen, Sourish Iyengar, Shila Ghazanfar, Ellis Patrick.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
